packages/core/src/network.js

```diff
@@ -8,16 +8,50 @@ const ALLOWED_STATUSES = [200, 201, 301, 302, 304, 307, 308];
 const ALLOWED_RESOURCES = ['Document', 'Stylesheet', 'Image', 'Media', 'Font', 'Other'];
 const ABORTED_MESSAGE = 'Request was aborted by browser';
 
+// DefaultMap, which returns a default value for an uninitialized key
+// Similar to defaultDict in python
+class DefaultMap extends Map {
+  constructor(getDefaultValue, ...mapConstructorArgs) {
+    super(...mapConstructorArgs);
+
+    if (typeof getDefaultValue !== 'function') {
+      throw new Error('getDefaultValue must be a function');
+    }
+
+    this.getDefaultValue = getDefaultValue;
+  }
+
+  get = (key) => {
+    if (!this.has(key)) {
+      this.set(key, this.getDefaultValue(key));
+    }
+
+    return super.get(key);
+  };
+};
+
+// RequestLifeCycleHandler handles life cycle of a requestId
+// Ideal flow:          requestWillBeSent -> requestPaused -> responseReceived -> loadingFinished / loadingFailed
+// ServiceWorker flow:  requestWillBeSent -> responseReceived -> loadingFinished / loadingFailed
+class RequestLifeCycleHandler {
+  constructor() {
+    this.releaseRequest = null;
+    this.releaseResponse = null;
+    this.checkRequestSent = new Promise((resolve) => (this.releaseRequest = resolve));
+    this.checkResponseSent = new Promise((resolve) => (this.releaseResponse = resolve));
```

Comment:
Can we try to make these names similar to actual lifecycle events ? so that its easier to understand. 

await requestWillBeSent; is better to read than await checkRequestSent imo. [ and reduces efforts needed to map it back to actual

Reply:
Sure, let name them as - `checkRequestWillBeSent` and `releaseRequestWillBeSent` ?

Reply:
`await requestWillBeSent` means "wait until `requestWillBeSent` event". So we dont need to add `check` prefix before it. In fact check makes it slightly confusing as it reads as "wait until checkRequestWillBeSent event". So my preference remains fro first [ even though its not blocking ] 

Similarly instead of `releaseRequestWillBeSent` would prefer `resolveRequestWillBeSent` as it means the same but one uses same word as in general used in code.